DVCodingStyle.md

```diff
@@ -461,6 +462,125 @@ utility macros for even more common code sequences.
         the end of the file, otherwise it will pollute the global namespace.
 
+#### Typical gotchas with macro usage:
+
+While it is better to avoid macros altogether, they do make our lives easy by
+shortening pieces of repeated code, in a manner that cannot be achieved with a
+function or a task. In general, the following describes the best practices to
+avoid common gotchas with macro usage.
+
+1.  Always wrap macro arguments **and** the output as applicable, within
+    parenthesis.
+
+    The following example illustrates how not wrapping the formal macro
+    arguments in parenthesis can cause unexpected behavior.
+    :-1:
+    ```systemverilog
+    `define AND(a, b) a && b
+
+    assign x = `AND(p || q, r) || s;
+    // Gotcha! wrong operator precedence! This results in: p || (q && r) || s which
+    // is different from what was originally intended: ((p || q) && r) || s.
+    ```
+    :+1:
+    ```systemverilog
+    `define AND(a, b) ((a) && (b))
+
+    assign x = `AND(p || q, r) || s;
+    // Correct: ((p || q) && (r)) || s
+    ```
+2.  Macro usage scope and avoiding namespace collisions.
+
+    *   If macros are defined for use in multiple sources, place them in a
+        separate SystemVerilog header file with appropriate `ifdef` guards.
+
+        This header file must not be compilation unit (provided as an input to
+        the tool for compilation). Instead, it should be `` `included`` in the
+        package, module or the interface that intends to invoke them.
+
+        Macro names should be appropriately prefixed with a string to avoid
+        potential namespace collisions. It is typical to use the file name for
+        the prefix string.
+
+        :-1:
+        ```systemverilog
+        `define STRINGIFY(s) `"s`"
+        // Gotcha! UVM library code also defines a macro with the same name.
+        // Depending on the tool, this may be flagged as an error. You may need
+        // this to support non-UVM testbench designs.
+        ```
+        :+1:
+        ```systemverilog
+        // file: dv_macros.h
+        `define DV_STRINGIFY(s) `"s`"
+        ```
+    *   If the scope of the macro usage is limited to the file, `undef` them
+        at the end of the file.
+
+        Macros have a global scope regardless of where they are defined. Some
+        tools do not even warn about namespace collisions, causing unnecessary
+        debug overhead. It is a good practice to `undef` macros at the end of
+        the file so that they are no longer visible to all subsequent
+        compilation units.
+3.  Avoid passing expressions as formal macro arguments.
+
+    As illustrated in the example below, it is better to avoid passing
+    expressions as arguments, since they are replaced verbatim. The macros may
+    reference the input arguments in multiple places, casuing the expressions to
+    be evaluated multiple times, resulting in unforeseen side effects.
```

Comment:
This only applies to side-effecting expressions. Since defining local variables is a bit of a bore in Verilog, let's not disallow things like `` `FOO(a+b)``.

Reply:
The point of `(`parenthesizing`)` the arguments is under the assumption that there will be expressions passed as arguments, so I think we need to allow for this.

Reply:
Removed.